py/dml/ctree.py

```diff
@@ -3958,8 +3958,16 @@ def mkCast(site, expr, new_type):
                 expr.site,
                 mkLit(expr.site, *real.get_store_fun()),
                 (mkCast(expr.site, expr, TInt(64, False)),))
+    if real.is_arith and old_type.is_arith:
+        assert (not (real.is_int and real.is_endian)
+                and not (old_type.is_int and old_type.is_endian))
+        return Cast(site, expr, new_type)
+    if ((isinstance(real, TPtr) or (real.is_int and not real.is_endian))
+        or (isinstance(old_type, TPtr)
+            or (old_type.is_int and not old_type.is_endian))):
```

Comment:
So this check is misleading -- endian ints already undergo conversion to regular ints earlier in the code. In fact, this check is morally incorrect because of this.
This should have the same asserts as the check before it, so you could combine the two: but that's messy.
```
    if ((real.is_arith and old_type.is_arith)
        or ((isinstance(real, TPtr) or real.is_int)
            and (isinstance(old_type, TPtr) or old_type.is_int))):
        assert (not (real.is_int and real.is_endian)
                and not (old_type.is_int and old_type.is_endian))
        return Cast(site, expr, new_type)
```
Any suggestions?

Reply:
The old check was incorrect in more than one way! (A mistaken `or` instead of `and`). When corrected, a couple of tests fail, all fixable without issue except for one: `1.2/misc/destroy`. This test does the following heinous crime:
```
method del_notify_cb(conf_object_t *subscriber, void *data) { ... }

typedef void (*notify_cb_t)(conf_object_t *subscriber, void *data);

method post_init {
  ...
    SIM_add_global_notifier(Sim_Global_Notify_Object_Delete, $dev.obj,
                            cast($del_notify_cb, notify_cb_t), NULL);
}
```
This is very much bad and wrong for multiple reasons -- for one, the generated function for `del_notify_cb` _has incorrect signature_ for the type `notify_cb_t` and so this test actually has undefined behavior.
This raises the very uncomfortable question if SSM abuses this, too. This just further proves we really need to test this against VP.